### PR TITLE
Disputes: Add Agreement dispute scripts 

### PR DESCRIPTION
**Pull request by Evalir**
**Merge after #2** 

Adds the scripts needed for the precedence campaign disputes.

- [x] WYSIWYG 1
- [x] WYSIWYG 2
- [x] WYSIWYG 3
- [x] WYSIWYG 4
- [x] WYSIWYG 5
- [x] 51% Attack
- [x] One Man Army

### CODE REVIEW

File: lib/web3-utils.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const RLP = require('rlp')
+
+// functions for counterfactual addresses
+async function buildNonceForAddress(_address, _index) {
```
Comment (facuspagnuolo) [inline lib/web3-utils.js]: Nit: are we using the `_` as a prefix for these JS helpers usually?

File: mainnet/14-wrong-permissions/evidence/against.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+The proposal Justification says:
```
Comment (facuspagnuolo) [inline mainnet/14-wrong-permissions/evidence/against.md]: Nit: should we prefix all the `wysiwyg` dispute dirs with that prefix? (similar to how we did for app mining)

File: mainnet/14-wrong-permissions/script.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+const {
+  buildNonceForAddress,
+  calculateNewProxyAddress,
+} = require('../../lib/web3-utils')
+
+module.exports = async () => {
+  // This variable is needed as if we leave the index param in the `buildNonceForAddress`
+  // function as 0, it will calculate the ACL address. For determining what number to assign,
+  // we need to count the number of apps installed, and then add 1.
+  const appIndex = 10.
```
Comment (facuspagnuolo) [inline mainnet/14-wrong-permissions/script.js]: Let's move this variable to one of the config parameters in the `script` object of the `metadata` file

File: mainnet/15-wysiwyg-wrong-permissions/evidence/against.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+The proposal Justification says:
+
+> This proposal will initialize the Execute actions permission on the Agent app, assign the Manager role to the Voting app, and assign the permission itself to the Tokens app.
+
+However the blockchain data shows:
+
+> ACL: Create a new permission granting `0x9958466922867D4fc0Da7451c9E717e4ff92D6fE` the ability to perform actions of role �Execute actions� on �Agent� (setting �Voting� as the permission manager)
```
Comment (facuspagnuolo) [inline mainnet/15-wysiwyg-wrong-permissions/evidence/against.md]:
```suggestion
> ACL: Create a new permission granting `0x9958466922867D4fc0Da7451c9E717e4ff92D6fE` the ability to perform actions of role "Execute actions" on "Agent" (setting "Voting" as the permission manager)
```

File: mainnet/14-wrong-permissions/evidence/against.md (hunk 2 above)
Comment (facuspagnuolo) [reply mainnet/14-wrong-permissions/evidence/against.md]: Second thought, this dir should probably be called `wrong-installation`

File: lib/encoder.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  const permissionsFunc = abi.encodeFunctionCall(aclAbi, [entity, counterfactualAppAddress, role, manager])
+  return [newAppInstanceFunc, permissionsFunc]
+}
+
+function encodeNewImmediatePayment(token, receiver, amount, reference) {
+  const newImmediatePaymentABI = getFunctionABI(FINANCE_ABI, 'newImmediatePayment')
+  return abi.encodeFunctionCall(newImmediatePaymentABI, [token, receiver, amount, reference])
+}
+
+function encodeWrongPermissionGrant(entity, app, role, manager) {
```
Comment (facuspagnuolo) [inline lib/encoder.js]: We can probably rename this one to `encodeCreatePermission` directly, even if we are using for a "malicious" intention

File: lib/encoder.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+function encodeCallsScript(actions) {
+  return actions.reduce((script, { to, data }) => {
+    const address = abi.encodeParameter('address', to)
+    const dataLength = abi.encodeParameter('uint256', (data.length - 2) / 2).toString('hex')
+    return script + address.slice(26) + dataLength.slice(58) + data.slice(2)
+  }, CALLSCRIPT_ID)
+}
+
+function encodeInvalidTransaction(ensContract, ensDomain, targetAgent) {
```
Comment (facuspagnuolo) [inline lib/encoder.js]: What about renaming this one to `encodeSetEnsOwnerThroughAgent` or similar?

**Review by facuspagnuolo — APPROVED** [review]
Looking awesome! Left some minor comments

File: lib/web3-utils.js (hunk 1 above)
Comment (Evalir) [reply lib/web3-utils.js]: Not really, my bad; forgot to change these when I moved them from a Discord conversation!

Comment (Evalir) [reply lib/web3-utils.js]: Done in https://github.com/aragon/court-precedence-campaign/pull/3/commits/3f4f8a23c7e71b4eeb561194ca4d221f98428425

File: mainnet/14-wrong-permissions/evidence/against.md (hunk 2 above)
Comment (Evalir) [reply mainnet/14-wrong-permissions/evidence/against.md]: Yep, let's do it so it's less confusing.

Comment (Evalir) [reply mainnet/14-wrong-permissions/evidence/against.md]: Done in https://github.com/aragon/court-precedence-campaign/pull/3/commits/d187dc1b0d51aa49674a62a8f3a24a855da0d9a7

File: mainnet/14-wrong-permissions/script.js (hunk 3 above)
Comment (Evalir) [reply mainnet/14-wrong-permissions/script.js]: Done in https://github.com/aragon/court-precedence-campaign/pull/3/commits/6969a24a5b1c132512ea0c05274b86d059795eb3

File: mainnet/15-wysiwyg-wrong-permissions/evidence/against.md (hunk 4 above)
Comment (Evalir) [reply mainnet/15-wysiwyg-wrong-permissions/evidence/against.md]: Done in https://github.com/aragon/court-precedence-campaign/pull/3/commits/f412eb172d8aee814638bfa7b243d393bb5b7134

File: lib/encoder.js (hunk 5 above)
Comment (Evalir) [reply lib/encoder.js]: Rename this functions and others in the `encoder.js` file for more clarity in this commit: https://github.com/aragon/court-precedence-campaign/pull/3/commits/d6bf4e2b628ad9a2539475a06c56829c0fcea297

- `encodeWrongPermissionGrant` should be `encodeCreatePermission`
- `encodeInvalidTransaction` should be `encodeSetEnsOwnerThroughAgent`
- `encodeMaliciousInstallation` should be `encodeNewAppInstance`

File: lib/encoder.js (hunk 6 above)
Comment (Evalir) [reply lib/encoder.js]: See https://github.com/aragon/court-precedence-campaign/pull/3#discussion_r480154593

File: mainnet/16-wysiwyg-invalid-agent-tx/evidence/for.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+Name: @light 
+
+Contact: @light in the Aragon Chat and Aragon Forum
+
+Summary: Transfers control of the `ancash.eth` ENS name from one of our Agent addresses to our other AN Cash Agent address.
+
+Rationale: We decided to use the AN Cash Agent address `0x87f4dc77707cdbedd0069c3cfc1b96f1725f03e0` to hold all of our ENS names.
+
+Description: If approved, this proposal will transfer the controller of the `ancash.eth` ENS name from AN Cash Agent address `0xF46Ac50c67E77Bc5CddDEf0E323963B6F0AEAde8` to the AN Cash Agent address `0x87f4dc77707cdbedd0069c3cfc1b96f1725f03e0`.
```
Comment (sohkai) [inline mainnet/16-wysiwyg-invalid-agent-tx/evidence/for.md:9]: Should we/this Agent be in control of this ENS name?

Comment (Evalir) [reply mainnet/16-wysiwyg-invalid-agent-tx/evidence/for.md:9]: The agent that will transfer the ENS name needs to be the controller of it, yes.

File: mainnet/13-wysiwyg-finance-transfer/evidence/against.md [hunk 8]
```diff
@@ -13,3 +13,4 @@ Given guideline 3.2 of the Aragon Network Cash Agreement:
 - Proposals MUST be accompanied by an accurate American English-language Justification of the proposal.
 
 This proposal should be ruled "BLOCK" by Aragon Court.
+
```
Comment (sohkai) [inline mainnet/13-wysiwyg-finance-transfer/evidence/against.md:16]: Just double checking, but should we remove this newline now then?

File: mainnet/14-wysiwyg-wrong-installation/script.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  encodeNewRewardsAppInstance,
+  encodeCallsScript,
+} = require('../../lib/encoder')
+const {
+  buildNonceForAddress,
+  calculateNewProxyAddress,
+} = require('../../lib/web3-utils')
+
+module.exports = async () => {
+  const nonce = await buildNonceForAddress(voting, appIndex)
```
Comment (sohkai) [inline mainnet/14-wysiwyg-wrong-installation/script.js:28]: Ah, there's a misunderstanding here.

You don't need `appIndex`—you should be able to do this like this:

```
await buildNonceForAddress(kernel, 0)
```

**Pull request merged**